presto-main/src/main/java/io/prestosql/operator/PageBuffer.java

```diff
@@ -34,10 +38,25 @@
     @Nullable
     private Page page;
     private boolean finished;
+    private Optional<ListenableFuture<?>> initialFuture;
+
+    public PageBuffer()
+    {
+        this.initialFuture = Optional.empty();
+    }
+
+    public PageBuffer(ListenableFuture<?> initialFuture)
+    {
+        this.initialFuture = Optional.of(requireNonNull(initialFuture, "initialFuture is null"));
+    }
 
     public WorkProcessor<Page> pages()
     {
         return WorkProcessor.create(() -> {
+            if (initialFuture.isPresent() && !initialFuture.get().isDone()) {
```

Comment:
make it non-Optional, and initialize it to completed future when one is not provided